File: apps/canopy_sessions/page_state.py
"""What the user is looking at, as state the agent can re-read.

The sibling of `page_actions`: that module is what the page can DO, this is
what the page currently SHOWS. Together they are the two halves AG-UI splits a
frontend into — frontend tools and shared state — and the second half is the
one canopy did not have.

**Why this is not "context appended to the prompt".** Before this, the page
snapshot was fetched once when the frame initialised, rendered to prose, and
pasted onto the first message. Three things were wrong with that and only the
third is obvious:

  * it was read at the wrong MOMENT — the user filters the page *after* opening
    the widget far more often than before, so the one snapshot we took was
    routinely of a view that no longer existed;
  * it reached the agent as PROSE, which is measurably worse than structure at
    equal size (the WebMCP origin-trial ablation recovered the right next action
    19/20 from a bounded structured packet versus 14/20 from a byte-matched
    free-text summary);
  * it was attached to the FIRST message only, so by turn three the agent was
    reasoning about a page the user had left.

State fixes all three by being a thing the agent can read *whenever it asks*,
rather than a thing the page says once.

**It is a cache, never the authority.** Exactly as with contacts. The state says
*which* rows are on screen; the agent re-reads the rows themselves through the
ordinary MCP tools, where the caller's own ACL applies. That is why the size cap
below is deliberately too small to hold the rows: a page that tries to send its
data instead of its selection is refused, out loud, at the point the mistake is
made rather than in a code review nobody runs.
"""

from __future__ import annotations

import json

from .models import Session

#: The whole point is a BOUNDED packet. 8 KiB holds a filter set and several
#: hundred ids comfortably and cannot hold the rows behind them, which is the
#: constraint we actually want to enforce: send the selection, not the data.
#: Chosen to be generous for the intended shape and hostile to the wrong one.
MAX_STATE_BYTES = 8192

#: Reserved: the tool a page names as backing its visible rows. Not validated
#: against the MCP registry here (this app must not import `apps.mcp`), but
#: carried through so the agent is told where to go and read properly.
BACKING_TOOL_KEY = "backing_tool"


class PageStateError(Exception):
    """A declaration the page must fix, with a `code` the caller can branch on.

    `too_large` is the only one that fires in practice, and it is a design
    guard rather than a resource limit — see MAX_STATE_BYTES.
    """

    def __init__(self, code: str, message: str):
        self.code, self.message = code, message
        super().__init__(message)


def current_page_state(session: Session) -> dict:
    """What the attached page currently shows.

    An empty dict means no page is attached or the page declares nothing — NOT
    that the page is empty. A caller that cannot tell those apart will describe
    a blank screen to the user.
    """
    return dict(session.page_state or {})


def set_page_state(session: Session, state: dict) -> dict:
    """Replace the declaration wholesale, and bump the version.

    Never merged, for the same reason actions are not: a page has ONE current
    view, and a key left over from the page the user navigated away from is a
    key the agent would reason about as if it were still on screen. Merging
    makes stale state indistinguishable from fresh state, which is the failure
    this whole module exists to remove.

    The version is monotonic per session so a consumer can tell one snapshot
    from the next, and so a later delta-based transport has an ordering to apply
    against. It is assigned here rather than accepted from the page: a client
    that picks its own version numbers can silently go backwards.
    """
    if not isinstance(state, dict):
        raise PageStateError("bad_state", "page state must be a JSON object")

    encoded = json.dumps(state, separators=(",", ":"))
    if len(encoded.encode("utf-8")) > MAX_STATE_BYTES:
        raise PageStateError(
            "too_large",
            f"page state is larger than {MAX_STATE_BYTES} bytes. Send the "
            f"selection (ids, filters) and the tool that resolves it, not the "
            f"rows themselves — the agent reads those itself, with the user's "
            f"own permissions applied.",
        )

    previous = dict(session.page_state or {})
    version = int(previous.get("version") or 0) + 1
    # `version` lives inside the blob rather than in its own column: it has no
    # meaning apart from the state it stamps, and a separate column invites the
    # two to be written apart and disagree.
    session.page_state = {**state, "version": version}
    session.save(update_fields=["page_state", "updated_at"])
    return dict(session.page_state)


def clear_page_state(session: Session) -> None:
    """Forget the view. Called when a page detaches.

    Clearing is not the same as declaring `{}`: both read as "nothing on
    screen", which is correct, and neither is the same as never having spoken.
    """
    if session.page_state:
        session.page_state = {}
        session.save(update_fields=["page_state", "updated_at"])

File: frontend/src/widget/pageState.ts
/**
 * The page's side of the state channel.
 *
 * `pageActions` says what the page can DO; this says what it currently SHOWS.
 * They are the two halves AG-UI splits a frontend into — frontend tools and
 * shared state — and this is the half canopy did not have.
 *
 * **Why this replaces `pageContext`'s snapshot.** `buildPageContext` was read
 * once, when the frame initialised, and pasted onto the first message as prose.
 * A user filters the page *after* opening the chat at least as often as before,
 * so the one reading we took was routinely of a view that no longer existed,
 * and nothing updated it for the rest of the conversation. State is pushed on
 * every change instead, exactly like the action registry, so the agent's view
 * of the screen cannot silently drift from the screen.
 *
 * A module-level registry for the same reason the action one is: the consumer
 * is not a React component. The widget mounts its own DOM outside the tree.
 */

/** What a page declares. Deliberately open — canopy does not know what any
 *  given host's page means — but see `describeSelection` for the shape that
 *  actually works, and `MAX_STATE_BYTES` on the server for the bound that
 *  stops a page sending its rows instead of its selection. */
export type PageState = Record<string, unknown>

type Contributor = () => PageState

let contributor: Contributor | null = null
const listeners = new Set<(state: PageState) => void>()

/** Assemble the current state, never throwing.
 *
 *  A contributor that throws must cost only its own contribution: the route
 *  layer beneath it is still true, and an agent with the path and no selection
 *  is far better off than an agent with nothing.
 */
export function currentPageState(base: PageState = {}): PageState {
  if (!contributor) return { ...base }
  try {
    return { ...base, ...contributor() }
  } catch {
    return { ...base }
  }
}

/** Whether any page has declared a view.
 *
 *  Distinct from `currentPageState()` returning `{}`: a page that declares
 *  nothing must not cause the widget to announce the user's screen as empty,
 *  which would overwrite a perfectly good declaration with an assertion that
 *  there is nothing to see.
 */
export function hasPageState(): boolean {
  return contributor !== null
}

/** Register the page's contributor. Last wins; returns a disposer. */
export function setPageStateContributor(fn: Contributor): () => void {
  contributor = fn
  notifyPageStateChanged()
  return () => {
    // Only if it is still OURS: React can run an old component's cleanup after
    // a new one registered, and clearing unconditionally would blank the state
    // of the page the user is actually on.
    if (contributor === fn) {
      contributor = null
      notifyPageStateChanged()
    }
  }
}

/** Tell the widget the view moved. Called by `usePageState` on every change. */
export function notifyPageStateChanged(): void {
  const state = currentPageState()
  listeners.forEach((l) => l(state))
}

export function onPageStateChanged(fn: (state: PageState) => void): () => void {
  listeners.add(fn)
  return () => listeners.delete(fn)
}

/**
 * The shape that actually works, as a named helper rather than a doc comment.
 *
 * The rule it encodes: **send the selection, not the data.** `ids` says which
 * rows are on screen; `backingTool` says which MCP tool resolves them. The
 * agent then reads the rows through that tool, live, with the user's own
 * permissions applied — instead of trusting a copy the page serialised, which
 * can go stale between render and send and is a second place an ACL could be
 * got wrong.
 *
 * This is the same rule contacts follow: the local copy is a cache, never the
 * authority.
 */
export function describeSelection(input: {
  /** The MCP tool that resolves these ids. */
  backingTool: string
  /** The MCP resource URI this page is showing, e.g. `insight://`.
   *
   *  This is what canopy keys INVALIDATION on: when the data behind it changes
   *  — by any actor, through any door — every page declaring this resource is
   *  told to re-read. A page that omits it still describes itself to the agent
   *  but will never be told its data moved, which is a worse page rather than a
   *  broken one. */
  resource?: string
  /** What is on screen, in display order. */
  ids: Array<string | number>
  /** The filters producing that selection, if they are not already in the URL. */
  filters?: Record<string, unknown>
  /** Anything else about the view that is not derivable from the URL. */
  extra?: Record<string, unknown>
}): PageState {
  const { backingTool, resource, ids, filters, extra } = input
  return {
    backing_tool: backingTool,
    ...(resource ? { resource } : {}),
    visible_ids: ids,
    visible_count: ids.length,
    ...(filters && Object.keys(filters).length ? { filters } : {}),
    ...extra,
  }
}
